### PR TITLE
Fix lineinfo from wat2wasm

### PR DESCRIPTION
**Pull request by tolauwae**
Fixes an issue for large `.wast` files generated by AssemblyScript, where `wat2wasm` gave different (wrong) line info than `wasmobjdump`.

Issue didn't manifest for small handwritten files.

**The underlying issue:** wat2wasm is a two pass assembler. We can only print line info in the first pass, where the size of the function only takes in one byte. However, large function can use two (or more) bytes in the second pass.

**The fix:** by passing the `--no-canonicalize-leb128s` option we always use 5 bytes now. We also need to disable bulk memory, for an unknown reason this section can also change the addresses of the instructions between passes.

The `--no-canonicalize-leb128s` option increases binary size, so ideally we would remove it and get the line info in a different way.

### CODE REVIEW

**Comment by carllocos** [comment]
I also experienced a similar issue and indeed I only experienced it when compiling relatively large Wast files to wasm. In fact. I discovered that `wat2wasm` printed wrong hexa addresses for any wasm section that becomes greater than 256 bytes (so not only for code sections but also for import section, table section, etc.). So in my case, the line numbers were not incorrect but the hexa addresses of a section larger than 256 bytes were. And that makes the plugin fail to retrieve the right line number based on the pc since the plugin works with incorrect hexa addresses.

I did not know it was possible to fix this issue at the compiler level as you are trying to do above. So in my case, I added additional code that fixes the problem e3f4c02. My solution is certainly not the right one and should be removed in favour of the one above. And it could also be the reason why your solution may not work.

**Comment by tolauwae** [comment]
Ok than why wait? just git revert e3f4c02 and check if it works

**Comment by carllocos** [comment]
I do not even get the chance to test without the commit because the compilation crashes before I event get to the piece of code I added. I get this error

```
HardwareDebugBridge stderr: cp: cannot stat '/tmp/warduino.BcUDC7/upload.wasm': No such file or directory
make: *** [Makefile:23: bin/upload.h] Error 1
```

**Comment by tolauwae** [comment]
I undid the changes. They also happened to break the compile test in the plugin, which is nice to see the CI working.

@carllocos ready for review

**Review by carllocos — COMMENTED** [review]
@tolauwae depending on the wast example that I try to debug the plugin fails to open. I get this error `EmulatedDebugBridge stderr: cannot open file`.

Logging the `stderr` of `handleStdError` method from the WASMCompilerBridge I get the following output:
- `examples/temp_monitor.wast:36:3: error: undefined table variable \"$e0\"\n (elem $e0 (i32.const 0) $disconnectSensors)\n  ^^^^\n`
-  `examples/dimmer-m5stickc.wast:113:4: error: undefined table variable \"$e0\"\n  (elem $e0 (i32.const 1) func $decrease)\n`

This PR works fine for *examples/blink.wast*. So it may just be that the examples for which the plugin fail are simply wrong. The error seems to be always the same `undefined table variable`

**Comment by tolauwae** [comment]
Like I said before, the `--disable bulk-memory` option is the reason `wat2wasm` cannot compile these examples anymore.
So the examples are not wrong, they just use the bulk memory proposal of Wasm.
Apparently table variables is something that is included in this proposal, but I haven't double checked this yet.

But since WARDuino doesn't actually support this proposal, this is not actually a problem for the plugin technically.

So I don't know if we want to stall this PR until we fix the examples or we look at this is a next one?
We can certainly merge this PR imo, we should just fix the examples before we release version 0.4.1.

**FYI** In AssemblyScript you can easily disable the proposal as well, so all the generated code should work.
I haven't found a way in Rust.

**Comment by carllocos** [comment]
I'm not sure to understand your answer correctly. Does it mean that we cannot write any wast code that makes use of the table entries? Because in that case it may deactivate the use of callback system in WARDUino. If I recall correctly, the callbacksystem uses the table references to know which functions are callbacks. Registering a callback happens via `subscribe_interrupt` primitive where one of the arguments is the table index that points to the callback function. If the only way to populate the table entries is via `elem` e.g., `(elem $e0 (i32.const 0) $disconnectSensors)`. 

Regarding this:
> So I don't know if we want to stall this PR until we fix the examples or we look at this is in a separate PR?

If my concern here above is valid then I think we should delay. Otherwise, I can accept the PR.

**Comment by tolauwae** [comment]
> I'm not sure to understand your answer correctly.

TLDR: The problem is caused by the `--disable bulk-memory` option for `wat2wasm`. The WARDuino VM does not support bulk memory, so neither does the plugin need to.

> Does it mean that we cannot write any wast code that makes use of the table entries?

No, it means we cannot use table variables, but tables work.

This does not work:

```wasm
 (table $0 2 funcref)
 (elem $e0 (i32.const 1) $hardware/buttonPressed)
```

This does work:

```wasm
 (table $0 2 funcref)
 (elem (i32.const 1) $hardware/buttonPressed)
```

> If I recall correctly, the callbacksystem uses the table references to know which functions are callbacks.

Yes. This still works without bulk memory.

**Review by carllocos — APPROVED** [review]
In that case, the PR seems fine. I tested on my side and it still works for me.
